src/buffer.rs

```diff
@@ -775,6 +775,14 @@ where
         *self.get_pixel(x, y)
     }
 
+    fn opt_pixel(&self, x: u32, y: u32) -> Option<P> {
+        if let Some(idx) = self.pixel_indices(x, y) {
+            Some(*P::from_slice(&self.data[idx]))
```

Comment:
@HeroicKatora : Is the bound check on `self.data` correctly eliminated here ?

Reply:
It would require the optimizer to note that the length of the slice is guaranteed to be longer than the maximum pixel index. I'm not entirely sure it does or can. I feel slightly uncomfortable replacing it with an `unsafe` version since the `ImageBuffer` is parameterized over a user provided buffer but a fallible indexing could avoid the panic path and instead have a `cmov` that is never taken. That might vectorize better than a panic path.

Reply:
I don't think it's possible to create an `ImageBuffer` where the backing buffer is too short, is it ? This function is likely to be called in a tight loop in a performance-sensitive part of a program, I think using a single line of unsafe should be worth the performance gains.

Reply:
It's certainly not possible to create a `ImageBuffer<_, &mut [u8]>` where the backing buffer is too short but the buffer offers customization of the container type. In those cases, there is nothing preventing an incorrectly implemented container from lying to the constructor to fake the correct length and subsequently shortening its internal length. It wouldn't even be `unsafe` to do so.

I think to fully use `unsafe` for this access we'd need to create a small wrapper type that can only be constructed with a buffer that is one of the standard types. Similar to `Pin<&mut T>` this could be constructed from any `ImageBuffer` by dereferencing the container.

Reply:
Yes, this would be better, but this would be specific to the *standard* types. I was thinking of something that would still work for all `Container: Deref<Target = [P::Subpixel]>` and yet prevent the duplicated bound checks: we could check only whether `x < width`, then compute the pixel indices, and then use [`get`](https://doc.rust-lang.org/std/primitive.slice.html#method.get) on the slice (which return an option, that would be None if `y < height`). Would that work ?

Reply:
Yep, the option trick would work for all types. The `get_unchecked` trick that completely eliminates the bounds check only works for standard types.